### PR TITLE
h2m/m2h: l10n support for cards

### PR DESCRIPTION
**Pull request by SphinxKnight**
This PR is an attempt at helping https://github.com/mdn/translated-content/discussions/1236 for localizers in general and https://github.com/mdn/translated-content/pull/1574 for French in particular.

Technically this handles div.note / div.warning and div.callouts elements.

cc @Gregoor

### CODE REVIEW

**Comment by SphinxKnight** [comment]
A few notes: 
* I went for gettext with JSON as it's still simple enough and seems a good enough foundation if other files are added / Pontoon somehow gets connected with yari
* To test this, one can run `yarn run md h2m web/javascript --locale fr --mode dry` before (## Top 20 unhandled elements - div.note (134) ) and after (## Top 20 unhandled elements - tr (133))

**Comment by SphinxKnight** [comment]
In terms of logic, this:
* convey/propagate the locale option from https://github.com/mdn/yari/pull/4026 to h2m (previously this just restricted the scope of files to the locale)
* uses this option in the card handler
* adds node-gettext as a new dependency (which I chose because it looked simple enough (not a whole l10n framework) and used gettext which is quite versatile / largely used. The JSON format can easily be converted to .po files if one decides to change l10n stuff.

**Comment by SphinxKnight** [comment]
\<aside\>well, it seems you cannot change the name of a branch afterwards\<\\aside\> Thanks to @hochan222, this PR is not restricted to fr anymore :)

**Comment by SphinxKnight** [comment]
@fiji-flo is it possible to have a look on this so we can get locales on par with en-US (cf. https://github.com/mdn/content/pull/7092)

File: markdown/cli.ts [hunk 1]
```diff
@@ -164,8 +164,11 @@ program
             console.log(doc.metadata.slug);
           }
           const { body: h, attributes: metadata } = fm(doc.rawContent);
+          const concreteLocale =
+            options.locale === "all" ? "en-US" : options.locale;
```
Comment (fiji-flo) [inline markdown/cli.ts:168]: there should be `doc.locale` available. And this way it even works properly when building all locales.

File: markdown/h2m/handlers/cards.ts [hunk 2]
```diff
@@ -1,13 +1,26 @@
+import * as Gettext from "node-gettext";
 import { h } from "../h";
 import { asArray } from "../utils";
 import { toText } from "./to-text";
 import { QueryAndTransform } from "./utils";
 
+const getTextDefaultDomainName = "messages";
+
+function initLocalizations(locale) {
```
Comment (fiji-flo) [inline markdown/h2m/handlers/cards.ts:9]: Could we initialize these only once? Either lazy of all locales at once.

**Review by fiji-flo — CHANGES_REQUESTED** [review]
Thanks for starting this. With some changes this will look good. But the much more important markdown → html conversion part is missing (https://github.com/mdn/yari/blob/main/markdown/m2h/handlers/index.js#L5)

File: markdown/cli.ts (hunk 1 above)
Comment (SphinxKnight) [reply markdown/cli.ts]: Thanks, it looks better now :)

File: markdown/h2m/handlers/cards.ts (hunk 2 above)
Comment (SphinxKnight) [reply markdown/h2m/handlers/cards.ts]: I was not fully sure of what you expected and I went with https://github.com/mdn/yari/pull/4251/commits/e9cccd723a9793471befb16730d23c24d7ab926e, let me know if it's OK.

**Comment by SphinxKnight** [comment]
I'll work on the m2h side as well. Just out of curiosity, I thought this was just intended as a manual tool: will the Markdown to HTML conversion be used server side?

**Comment by SphinxKnight** [comment]
sorry for the push, I had forgotten to sign my commits.

File: markdown/cli.ts [hunk 3]
```diff
@@ -196,16 +201,23 @@ program
   )
 
   .command("m2h", "Convert Markdown to HTML")
+  .option("--locale", "Targets a specific locale", {
```
Comment (SphinxKnight) [inline markdown/cli.ts:204]: I added the locale option here as well as it was far easier to test on a specific locale.

File: markdown/m2h/index.js [hunk 4]
```diff
@@ -6,32 +6,36 @@ const gfm = require("remark-gfm");
 const raw = require("rehype-raw");
 const format = require("rehype-format");
 
-const handlers = require("./handlers");
+const buildLocalizedHandlers = require("./handlers");
 const { decodeKS, encodeKS } = require("../utils");
 
-function makeProcessor() {
+function makeProcessor(options) {
+  const localizedHandlers = buildLocalizedHandlers(options.locale);
```
Comment (SphinxKnight) [inline markdown/m2h/index.js:13]: I wasn't able to find a more elegant way to convey the option/locale to the handler so I went with "building" them here.
Let me know if you think of something which could be better.

**Comment by SphinxKnight** [comment]
(rebasing in order to get changes from #4259 )

**Comment by fiji-flo** [comment]
I just glanced through this, looks good so far. But I need to really review this. I'm on vacation right now but this is top of my list next week when I'm back! 💯

**Comment by SphinxKnight** [comment]
> 
> 
> I just glanced through this, looks good so far. But I need to really review this. I'm on vacation right now but this is top of my list next week when I'm back! 💯

Thanks for the message/update: enjoy your vacation :) there'll be plenty of time to see that afterwards:)

**Comment by fiji-flo** [comment]
@SphinxKnight Sorry for the delay. But I think I was lacking context when I first looked at it.

I don't know but I was under the impression that this is content related as in these changes will be display in the rendered HTML. However, this is purely MDN specific markdown syntax as this keywords are stripped out when transforming markdown to HTML.

I do see the point that the markdown is easier to read this way, however it comes with some complexity. Can I have some thoughts on that @wbamberg @escattone

**Comment by SphinxKnight** [comment]
As a localizer, the intent of this PR is:
* to have the h2m conversion tool to be happy with the current localized keywords in the translated content (technically: having less conversion errors/unhandled elements based on the existing localized content which should already use the same keyword for each card type)
* to be able to have consistant localized content while editing in markdown and to avoid mixing english keywords both in the markup/translated docs with the localized text (technically, not having to fallback to writing inline HTML in markdown docs just to have "Attention :" instead of "Warning:")

I proposed this PR to tackle those two points which I think are necessary for l10n.
If there is anything unclear, let me know.

File: markdown/h2m/handlers/cards.ts [hunk 5]
```diff
@@ -1,13 +1,35 @@
...
+})();
+
 export const cards: QueryAndTransform[] = [
   ...["note", "warning"].map(
     (className) =>
       [
-        (node) => {
+        (node, opts) => {
+          const locale = opts.locale || DEFAULT_LOCALE;
+          const gt = gettextLocalizationMap.get(locale);
```
Comment (fiji-flo) [inline markdown/h2m/handlers/cards.ts:32]: We should default this to English.

**Review by fiji-flo — CHANGES_REQUESTED** [review]
Could you add a fallback to English keywords? Everything else looks good!

File: markdown/h2m/handlers/cards.ts [hunk 6]
```diff
@@ -1,13 +1,35 @@
+import * as Gettext from "node-gettext";
+import * as fs from "fs";
+import * as path from "path";
+import { DEFAULT_LOCALE } from "../../../libs/constants";
 import { h } from "../h";
 import { asArray } from "../utils";
 import { toText } from "./to-text";
 import { QueryAndTransform } from "./utils";
 
+const gettextLocalizationMap = (function initLocalizationsMap() {
```
[the remark this answers is not in the document]
Comment (SphinxKnight) [reply markdown/h2m/handlers/cards.ts]: none, fixed :)

File: markdown/h2m/handlers/cards.ts (hunk 5 above)
Comment (SphinxKnight) [reply markdown/h2m/handlers/cards.ts:31]: I'm not sure if I get your intent right but please see 5239908

Comment (fiji-flo) [reply markdown/h2m/handlers/cards.ts:31]: @SphinxKnight 
Sorry that I didn't make myself clear, it's about `yarn md h2m web/html --locale de` not failing. We should default to the en-US keywords if there's on localization.

I'd even argue that the en-US keywords should always work even if it has been localized.

Comment (SphinxKnight) [reply markdown/h2m/handlers/cards.ts]: agreed, I should have seen this one.
Thanks for the review, I've pushed cd26305 and hope it's OK.

**Review by fiji-flo — APPROVED** [review]
Code wise this looks okay. @wbamberg could you take a look from the content perspective, then we can merge this.

**Comment by wbamberg** [comment]
Thanks for this @SphinxKnight ! I'm not going to try to understand all the code, but just see how it looks and behaves.

One thing is, it looks like this is regressing callouts. Look at, say, http://developer.mozilla.org/en-US/docs/Web/CSS . On MDN (and `yarn dev` from mdn/yari/main) we get a callout:

Screen Shot 2021-08-31 at 9 19 03 AM

Running this branch I get:

Screen Shot 2021-08-31 at 9 22 31 AM

...so it's getting "Warning" styling, and the "Callout:" magic word is not getting snipped out.

**Comment by wbamberg** [comment]
Also if I edit a note so the first element is not a magic word, but is still `<strong>`:

```
> **Notanote:** For documents whose [root element](/en-US/docs/Web/HTML/Element/html) is an HTML element...
```

...then in mdn/yari/main I get the expected:

Screen Shot 2021-08-31 at 9 29 43 AM

....but in this branch I get a page build error:

```
Error: 500 on /en-US/docs/Web/CSS/background-clip/index.json: TypeError: Cannot read property 'split' of undefined
```

File: markdown/localizations/en-US.json [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+      "card_note_label": {
+        "msgid": "card_note_label",
+        "msgstr": ["Note:"]
+      },
+      "card_warning_label": {
+        "msgid": "card_warning_label",
+        "msgstr": ["Warning:"]
+      },
+      "card_callout_label": {
+        "msgid": "card_warning_label",
```
Comment (fiji-flo) [inline markdown/localizations/en-US.json:13]: @SphinxKnight @wbamberg that's the reason for the broken callout (sorry I missed that). And it's happening in other locales, too.

**Review by fiji-flo — CHANGES_REQUESTED** [review]
@wbamberg Thanks for taking a proper look. It's my last day working for a month which is a bad excuse.

@SphinxKnight please do address the resilience issues.

**Comment by wbamberg** [comment]
It would be great to have automated regression tests for m2h at least.

File: markdown/localizations/en-US.json (hunk 7 above)
Comment (SphinxKnight) [reply markdown/localizations/en-US.json]: Thanks to both of you, this one has been taken care of w/ 7e66a5f

**Comment by SphinxKnight** [comment]
I've fixed both issues, thanks again for your time looking at this.
@wbamberg could you confirm it's now OK for you? :x

**Comment by wbamberg** [comment]
> I've fixed both issues, thanks again for your time looking at this.
> @wbamberg could you confirm it's now OK for you? :x

Yes, it looks good to me now. Thanks for the fixes @SphinxKnight !

**Comment by wbamberg** [comment]
@escattone , if Florian is out for a month it would be great to get someone to look at this PR. AIUI this is blocking other locales migrating to Markdown.

**Review by fiji-flo — APPROVED** [review]
I'm merging and deploying to dev. Please check out https://main.content.dev.mdn.mozit.cloud/en-US/docs/Web if you see any troubles.

**Pull request merged**